### PR TITLE
qemu: add annotations for iommu_platform for s390x virtio devices

### PR DESCRIPTION
**Pull request by huoqifeng**
qemu: Add iommu annotations for qemu for s390x ccw, so that the supported devices can make use of that.
for virtcontainers/hypervisor

add iommu_platform=on option for virtio devices

Fixes https://github.com/kata-containers/runtime/issues/2830

Signed-off-by: Qi Feng Huo <huoqif@cn.ibm.com>

### CODE REVIEW

**Comment by bpradipt** [comment]
@huoqifeng does this not help - https://github.com/kata-containers/runtime/blob/master/virtcontainers/pkg/annotations/annotations.go#L155 ?

**Comment by huoqifeng** [comment]
> @huoqifeng does this not help - https://github.com/kata-containers/runtime/blob/master/virtcontainers/pkg/annotations/annotations.go#L155 ?

oh, right It's exactly what I want. seems no code for ccw for s390x yet, I'll change the code and description of this PR. @bpradipt thanks for the info!

**Comment by huoqifeng** [comment]
https://github.com/kata-containers/runtime/blob/master/virtcontainers/qemu_s390x.go#L272, gonna implement here.

**Comment by huoqifeng** [comment]
oh, no, I just realized what I need is IOMMU_PLATFORM=on for virtio devices.

**Comment by c3d** [comment]
I don't see a configuration, only an annotation. Isn't that something that one might want to set as "on" by default in a configuration file?

**Comment by huoqifeng** [comment]
I'm actually struggling on this, is configuration needed to tune the default value? yeah, I can add it.

**Comment by ttx** [comment]
@huoqifeng we currently have false negative results for the SoB and WIP checks. I'm working on a fix, thanks for your patience.

File: vendor/github.com/intel/govmm/qemu/qemu.go [hunk 1]
```diff
@@ -385,6 +385,9 @@ func (fsdev FSDevice) QemuParams(config *Config) []string {
 		deviceParams = append(deviceParams, fmt.Sprintf(",romfile=%s", fsdev.ROMFile))
 	}
 	if fsdev.Transport.isVirtioCCW(config) {
+		if config.Knobs.IOMMUPlatform {
```
Comment (amshinde) [inline vendor/github.com/intel/govmm/qemu/qemu.go:388]: @huoqifeng Please submit this change to https://github.com/intel/govmm

**Comment by amshinde** [comment]
@devimc PTAL

File: virtcontainers/qemu.go [hunk 2]
```diff
@@ -26,7 +26,7 @@ import (
 
 	govmmQemu "github.com/intel/govmm/qemu"
 	"github.com/opencontainers/selinux/go-selinux/label"
-	"github.com/opentracing/opentracing-go"
+	opentracing "github.com/opentracing/opentracing-go"
```
Comment (devimc) [inline virtcontainers/qemu.go:29]: uhm this looks odd

**Comment by devimc** [comment]
cc @alicefr

File: vendor/github.com/intel/govmm/qemu/qemu.go (hunk 1 above)
Comment (huoqifeng) [reply vendor/github.com/intel/govmm/qemu/qemu.go:388]: Thanks for the comments, I'm current busy on other stuff, will check it and revise later...

File: virtcontainers/qemu.go (hunk 2 above)
Comment (huoqifeng) [reply virtcontainers/qemu.go:29]: Looks changed by my IDE go formatter. will change it back.

File: virtcontainers/pkg/oci/utils.go [hunk 3]
```diff
@@ -546,13 +555,13 @@ func addHypervisorMemoryOverrides(ocispec specs.Spec, sbConfig *vc.SandboxConfig
 		sbConfig.HypervisorConfig.HugePages = hugePages
 	}
 
-	if value, ok := ocispec.Annotations[vcAnnotations.IOMMU]; ok {
+	if value, ok := ocispec.Annotations[vcAnnotations.IOMMUPlatform]; ok {
```
[the remark this answers is not in the document]
Comment (huoqifeng) [reply virtcontainers/pkg/oci/utils.go:558]: oh, my fault.

[the remark this answers is not in the document]
Comment (huoqifeng) [reply virtcontainers/pkg/oci/utils.go:558]: Sure, will add.

**Comment by alicefr** [comment]
@devimc I'm leaving IBM and Z team at the end of the month, please ping @jschintag in the future. @huoqifeng My impression that the flag needs to be enabled first in [govmm](https://github.com/intel/govmm) as @amshinde already suggested. I'd also check before if the flag is used for pci devices and how.

File: vendor/github.com/intel/govmm/qemu/qemu.go (hunk 1 above)
Comment (jodh-intel) [reply vendor/github.com/intel/govmm/qemu/qemu.go:388]: Once the change lands in govmm, you then need to "re-vendor" it into this repo by creating a separate commit to the files changed below `vendor/`. That commit should list the changes in the upstream project. We've documented how to do this here:

https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#re-vendor-prs

It might sound complex but actually isn't :smile: Just ask us if unsure though.

Comment (huoqifeng) [reply vendor/github.com/intel/govmm/qemu/qemu.go:388]: PR https://github.com/intel/govmm/pull/132 created in repo govmm.

Comment (huoqifeng) [reply vendor/github.com/intel/govmm/qemu/qemu.go:388]: Created separate PR https://github.com/kata-containers/runtime/pull/2871 to revendor govmm, but it fails the travis check, the details is described in the new PR. @jodh-intel any suggestion?

Comment (jodh-intel) [reply vendor/github.com/intel/govmm/qemu/qemu.go:388]: Hi @huoqifeng - There are a few problems here, but don't worry they are all easy to resolve :smile:

First, if your PR needs to revendor one of the vendored packages like govmm, you can simply create a single commit as part of your main PR to do the revendor. By convention, the [subsystem](https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#subsystem) (which is the first token up to the colon on the first line of the commit) is "vendor".

As the link above shows, the subsystem *must* be a single word so that is one of reasons Travis is failing as you have specified two ("cf0f05d" and "qemu"):

```
cf0f05d qemu: add iommu_platform knob for qemuParams
```

The final issue is that the commit on #2871 doesn't have a body explaining the change in more details. We have a standard format for revendor commits: to list all the commits in that upstream project in the commit body (see `old_commit=...; new_commit=...` in this section: https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#re-vendor-prs

In summary, you need to do the following:

1) Add the revendor commit to this PR.
2) Make that revendor commit look something like this:

    ```
    vendor: Update govmm for s390x iommu_platform annoations

    Updated govmm to support iommu_platform annotations for virtio devices. govmm changes:

       cf0f05d qemu: add iommu_platform knob for qemuParams

    Signed-off-by: ...
    ```
3) Close #2871 since the change is now part of this PR.

----

See https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#patch-format for further details.

Comment (huoqifeng) [reply vendor/github.com/intel/govmm/qemu/qemu.go:388]: @jodh-intel thanks, it helps!

File: virtcontainers/qemu.go (hunk 2 above)
Comment (huoqifeng) [reply virtcontainers/qemu.go]: Fixed

File: virtcontainers/pkg/oci/utils.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (huoqifeng) [reply virtcontainers/pkg/oci/utils.go]: Fixed

File: cli/config/configuration-fc.toml.in [hunk 4]
```diff
@@ -136,6 +136,10 @@ block_device_driver = "@DEFBLOCKSTORAGEDRIVER_FC@"
 # command line: intel_iommu=on,iommu=pt
 #enable_iommu = true
 
+# Enable IOMMU_PLATFORM, default false
+# Enabling this will result in the VM device having iommu_platform=on set
```
[the remark this answers is not in the document]
Comment (huoqifeng) [reply cli/config/configuration-fc.toml.in:140]: Added.

**Comment by jodh-intel** [comment]
Thanks @huoqifeng, but the build is failing in Travis:

```
make: Entering directory '/home/travis/gopath/src/github.com/kata-containers/runtime/virtcontainers'
     GOBUILD  build
# github.com/kata-containers/runtime/vendor/github.com/intel/govmm/qemu
../vendor/github.com/intel/govmm/qemu/qemu.go:2484:17: config.Knobs.NoReboot undefined (type Knobs has no field or method NoReboot)
Makefile:32: recipe for target 'build' failed
```

Do you need to rebase against `master`?

**Comment by huoqifeng** [comment]
@jodh-intel I'm revdendering the govmm, it should fix the failure.

File: vendor/github.com/intel/govmm/qemu/qemu.go (hunk 1 above)
Comment (huoqifeng) [reply vendor/github.com/intel/govmm/qemu/qemu.go:409]: I suffered very slow `dep ensure` command, it's no help even I deleted Gopkg.lock file and the obsolete govmm vendor code.

**Comment by devimc** [comment]
@huoqifeng thanks, but looks like this pull request needs a rebase, you are trying to push 27 commits

**Comment by huoqifeng** [comment]
@devimc @jodh-intel Thanks for the help, I made progress on this, looks the travis job is good but there is another jenkins job always pending, any idea on this?

**Comment by devimc** [comment]
@huoqifeng thanks, I see two `Signed-off-by:` in your commit message, you should raise one commit for re-vendoring the packages and other commit for the fix.

**Comment by huoqifeng** [comment]
@devimc yes, amended the commits.

**Comment by devimc** [comment]
@huoqifeng please split this PR in two commits and I will help you to fix the CI, don't forget the `Signed-off-by:` in both commits

**Comment by huoqifeng** [comment]
@devimc The commit is split, would you please help on the CI failure?

**Comment by devimc** [comment]
@huoqifeng thanks, please fix your commit messages

commit  `vendor: Update govmm for s390x iommu_platform annoations`:
* Message length must be < 80 characters
* Missing < > in `signed-off` (Signed-off-by: Qi Feng Huo **\<huoqif@cn.ibm.com\>**)


commit `qemu: add annotations for iommu_platform for s390x virtio devices`:
* Missing body, for example:
```
qemu: add annotations for iommu_platform for s390x virtio devices

__ADD MORE DETAILS ABOUT THIS PATCH HERE___

Fixes #2830

Signed-off-by: Qi Feng Huo <huoqif@cn.ibm.com>
```

for more information about commits format see https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#patch-format

**Comment by huoqifeng** [comment]
> @huoqifeng thanks, please fix your commit messages
> 
> commit `vendor: Update govmm for s390x iommu_platform annoations`:
> 
> * Message length must be < 80 characters per line
> * Missing < > in `signed-off` (Signed-off-by: Qi Feng Huo **<[huoqif@cn.ibm.com](mailto:huoqif@cn.ibm.com)>**)
> 
> commit `qemu: add annotations for iommu_platform for s390x virtio devices`:
> 
> * Missing body, for example:
> 
> ```
> qemu: add annotations for iommu_platform for s390x virtio devices
> 
> __ADD MORE DETAILS ABOUT THIS PATCH HERE___
> 
> Fixes #2830
> 
> Signed-off-by: Qi Feng Huo <huoqif@cn.ibm.com>
> ```
> 
> for more information about commits format see https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#patch-format

Hope the amended commit message is OK.

**Comment by huoqifeng** [comment]
@jschintag @devimc I'm confused with the failure cases, any suggestions?

**Comment by jschintag** [comment]
ARM seems to be failling to install kubernetes, so that is unrelated to your PR. I have no expierience with clh so i cannot really say anything to those 2. The vfio CI is running `.ci/vfio_jenkins_job_build.sh` from the tests repository, haven't looked into that test either.

File: cli/config/configuration-fc.toml.in [hunk 5]
```diff
@@ -136,6 +136,11 @@ block_device_driver = "@DEFBLOCKSTORAGEDRIVER_FC@"
...
+# this option will add iommu_platform=true to the virtio PCI and CCW devices that support it 
+#enable_iommu_platform = true
```
Comment (sameo) [inline cli/config/configuration-fc.toml.in:142]: Are you sure this is relevant to Firecracker?

Comment (huoqifeng) [reply cli/config/configuration-fc.toml.in:142]: honestly, I'm not sure. I tested for qemu. It could be removed it if it's not a case firecracker support.

Comment (huoqifeng) [reply cli/config/configuration-fc.toml.in]: Removed it in fc configure.

**Comment by huoqifeng** [comment]
@amshinde @sameo are you happy with the changes?

**Comment by huoqifeng** [comment]
@sameo @jodh-intel @devimc Thanks for review and approve the code, looks the PR jobs are all good, may you please help merge it?

**Comment by jodh-intel** [comment]
Thanks again @huoqifeng.

**Comment by huoqifeng** [comment]
I'm happy as a kata user and as a contributor also now :-)

**Comment by jodh-intel** [comment]
:smile:
